### ISSUE TITLE
Docker mount $HOME/.config/path/file:/root/.config/path/file: succeeds mount, but no files on dest container

### ISSUE THREAD

**Issue by notjames** [docker/cli#5287]
### Description

Apologies for posting here. I posted in https://forums.docker.com/t/docker-mount-home-config-path-file-root-config-path-file-succeeds-mount-but-no-files-on-dest-container/142644 as well, but that forum appears to be on a deprecation path? If this is considered cross-posting, please forgive me. However, I do believe this is a bug, so I do not think this is cross-posting.

The following running container successfully mounts `/etc/ssl...`. However, on the running container, the `config.json` never exists while it does exist on the parent host. 

I've tried mounting the affected mount to various locations in the container and it never succeeds. Documentation suggested maybe running `sudo sysctl -w kernel.apparmor_restrict_unprivileged_userns=0`, which did not solve the problem. Moreover, neither use of `--privileged` nor creating the container as root via `sudo` solves the problem

### Reproduce

brand new installation of docker.io
```bash
apt install docker.io
```

as a non-priv user,
```bash
$ mkdir -p $HOME/.config/dynu
$ touch $HOME/.config/dynu/config
$ chmod 640 $HOME/.config/dynu/config
$ sudo chown $USER:docker $HOME/.config/dynu/config
```

then run
```bash
docker run -dp 80:80/tcp -p 443:443/tcp --network proxy-network \
  --add-host myserver:192.168.100.1 --name nginx-proxy \
  --mount type=bind,src="/etc/ssl/certs/nginx",target="/etc/ssl/certs/nginx"\
  --mount type=bind,src="$HOME/.config/dynu/config",target="/root/.config/dynu/config.json":ro \
  vproj-nginx:latest
```

YMMV obviously, but if memory serves, these steps should get you most of the way to reproduce the issue (provided it's a bug).

### Expected behavior

I expect that the path in the running container `/root/.config/dynu` exists and a file called `config.json` exists as readable only and contains the data of that on the host path.

### docker version

```bash
❯ docker version
Client:
 Version:           24.0.7
 API version:       1.43
 Go version:        go1.22.2
 Git commit:        24.0.7-0ubuntu4
 Built:             Wed Apr 17 20:08:25 2024
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          24.0.7
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.22.2
  Git commit:       24.0.7-0ubuntu4
  Built:            Wed Apr 17 20:08:25 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.12
  GitCommit:
 runc:
  Version:          1.1.12-0ubuntu3
  GitCommit:
 docker-init:
  Version:          0.19.0
  GitCommit:
```

### docker info

```bash
❯ docker info
Client:
 Version:    24.0.7
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  0.12.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx

Server:
 Containers: 27
  Running: 3
  Paused: 0
  Stopped: 24
 Images: 7
 Server Version: 24.0.7
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version:
 runc version:
 init version:
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.8.0-38-generic
 Operating System: Ubuntu 24.04 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 31.19GiB
 Name: jclinux
 ID: 50b1d456-89f5-4925-9461-dab8a2d1e5ec
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

I am using Ubuntu 24.04 LTS. This is a fresh re-install. I've tried volume mounts and bind mounts. Both give the same result.

I'm trying to migrate a container from my previous server. I rebuilt the image. When I attempt to run it with two mounts, one of the mounts does not actually mount anything. There are no errors from docker and the inspection shows that the mounts exist.

note that I have verified the file on the source-host side. It exists. 

```bash
 ❯ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 24.04 LTS
Release:        24.04
Codename:       noble
```

```bash
❯ uname -a
Linux myserver 6.8.0-38-generic #38-Ubuntu SMP PREEMPT_DYNAMIC Fri Jun  7 15:25:01 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by thaJeztah** [comment]
- What does `docker container inspect nginx-proxy` show? Does it have the right paths set? (Did your shell expand the `$HOME` env-var?)  
- And the daemon you're connected to is running locally (i.e. your `docker` cli is on the same host as the daemon is running?)


It's worth noting that docker 24.0.x reached EOL and is no longer maintained; I suspect you may have installed the distro-packaged version of docker; it's also worth trying to install docker's official packages; https://docs.docker.com/engine/install/ubuntu/

**Comment by rimelek** [comment]
@notjames 

> but that forum appears to be on a deprecation path?

I'm not sure what that means, but not relevant here, so after replying on the forum I just came to confirm that this command for example worked for me perfectly on an Ubuntu 24.04 tets VM with Docker CE 26.1.1.


```
docker run --rm -it --mount type=bind,src="$HOME/.config/test/config",target="/root/.config/test/config.json":ro bash
```

I haven't tested it with Docker CE 24.x, but it seems that there is no newer package for `docker.io` on Ubuntu 24.0.7. At least not on arm.

**Comment by notjames** [comment]
> * What does `docker container inspect nginx-proxy` show? Does it have the right paths set? (Did your shell expand the `$HOME` env-var?)
> 
>     * And the daemon you're connected to is running locally (i.e. your `docker` cli is on the same host as the daemon is running?)
> 
> 
> It's worth noting that docker 24.0.x reached EOL and is no longer maintained; I suspect you may have installed the distro-packaged version of docker; it's also worth trying to install docker's official packages; https://docs.docker.com/engine/install/ubuntu/

>  * What does `docker container inspect nginx-proxy` show? Does it have the right paths set? (Did your shell expand the `$HOME` env-var?)

Yes. It expanded $HOME correctly

>  * And the daemon you're connected to is running locally (i.e. your docker cli is on the same host as the daemon is running?)

Yes. The client and server are running on the same machine.

**Comment by notjames** [comment]
> @notjames
> 
> > but that forum appears to be on a deprecation path?
> 
> I'm not sure what that means, but not relevant here, so after replying on the forum I just came to confirm that this command for example worked for me perfectly on an Ubuntu 24.04 test VM with Docker CE 26.1.1.
> 
> ```
> docker run --rm -it --mount type=bind,src="$HOME/.config/test/config",target="/root/.config/test/config.json":ro bash
> ```
> 
> I haven't tested it with Docker CE 24.x, but it seems that there is no newer package for `docker.io` on Ubuntu 24.0.7. At least not on arm.

I doubt this is worth mentioning but for the sake of being thorough, this machine I'm using is x86_64

**Comment by notjames** [comment]
OK, I removed (purged) all old docker-ce and using the instruction linked above, installed the up-to-date docker. This one works! I had to obviously recreate the docker image and now as I think about this, I think the problem that was causing this is instead of running `docker exec -it ...`, I was running `docker run -it ...` -- sigh. feeling dumb but didn't notice it until this latest docker errored on my use of `run` 

My apologies for the false alarm.
